app/scripts/detect-multiple-instances.js

```diff
@@ -0,0 +1,55 @@
+/**
+ * Sets up two-way communication between the
+ * mainline version of extension and Flask build
+ * in order to detect & warn if there are two different
+ * versions running simultaneously.
+ */
+
+import browser from 'webextension-polyfill';
+import {
+  PLATFORM_CHROME,
+  PLATFORM_FIREFOX,
+  CHROME_BUILD_IDS,
+  FIREFOX_BUILD_IDS,
+} from '../../shared/constants/app';
+import { getPlatform } from './lib/util';
+
+const MESSAGE_TEXT = 'isRunning';
+
+const showWarning = () =>
+  console.warn('Warning! You have multiple instances of MetaMask running!');
+
+/**
+ * Handles the ping message sent from other extension.
+ * Displays console warning if it's active.
+ *
+ * @param message - The message received from the other extension
+ */
+export const onMessageReceived = (message) => {
+  if (message === MESSAGE_TEXT) {
+    showWarning();
+  }
+};
+
+/**
+ * Sends the ping message sent to other extensions to detect whether it's active or not.
+ */
+export const checkForMultipleVersionsRunning = async () => {
+  if (getPlatform() !== PLATFORM_CHROME && getPlatform() !== PLATFORM_FIREFOX) {
+    return;
+  }
+  const buildIds =
+    getPlatform() === PLATFORM_CHROME ? CHROME_BUILD_IDS : FIREFOX_BUILD_IDS;
+
+  const thisBuild = browser.runtime.id;
+
+  for (const id of buildIds) {
+    if (id !== thisBuild) {
+      try {
+        await browser.runtime.sendMessage(id, MESSAGE_TEXT);
+      } catch (error) {
```

Comment:
Hmm. I'm surprised the linter isn't complaining about this unused variable.

Reply:
In general, not checking the `error` type is a bad habit. We're assuming here that if this error is thrown, the receiving end does not exist. But we're not confirming that that's what this error is. It could be some unexpected error instead, indicating a bug somewhere in our code.

The practice I'd generally recommend is to test all expected flows, and ensure the thrown error is one of the expected ones. If not, re-throw it and we'll end up seeing it in Sentry and we can address it from there (either by fixing the bug or adding a new variant of expected error to the list).

Reply:
That's all totally reasonable. I did it this way to make this method secure from failing some other flows, since this functionality is pure informative, not operational. I think, adding logging of errors would be a useful improvement!

Reply:
Hmm. Logging could be useful, but I'm not sure about doing so indiscriminately. This is _expected_ to fail in most scenarios. Our startup logs are pretty concise right now so I don't think this would go unnoticed either. Also this message sounds like it's saying something went _wrong_.

Reply:
Yeah, you're right, that's probably too much logging. Ok, postponing this for now